logging/src/logging_v2.cpp

```diff
@@ -95,283 +93,344 @@ bool emit_ack(std::shared_ptr<logging::metafilter_t> filter,
-        list_scope(),
+        filter_unicorn_path(config.as_object().at("unicorn_path", "/cocaine/logging_v2/filters").as_string()),
         retry_timer(io_context)
     {
+        auto default_mf = get_default_metafilter();
+        auto default_metafilter_conf = config.as_object().at("default_metafilter").as_array();
+
+        std::map<std::string, dynamic_t> filter_info_conf;
+
+        // int64_t max value is used instead of uint64_t max value
+        // to prevent overflow in std::duration, from which timepoint is constructed.
+        filter_info_conf["deadline"] = static_cast<uint64_t>(std::numeric_limits<int64_t>::max());
+        filter_info_conf["logger_name"] = default_key;
+        filter_info_conf["disposition"] = static_cast<uint64_t>(disposition_t::local);
+        for(const auto& filter_conf : default_metafilter_conf) {
+            filter_info_conf["id"] = generate_id();
+            filter_info_conf["filter"] = filter_conf;
+            default_mf->add_filter(logging::filter_info_t(filter_info_conf));
+        }
+
+        auto core_mf = get_core_metafilter();
+        cocaine::filter_t core_filter([=](blackhole::severity_t sev, blackhole::attribute_pack& pack) {
+            logging::filter_result_t result;
+            if(core_mf->empty()) {
+                result = default_mf->apply(sev, pack);
+            } else {
+                result = core_mf->apply(sev, pack);
+            }
+            if(result == logging::filter_result_t::accept) {
+                return true;
+            } else {
+                return false;
+            }
+        });
+        context.logger_filter(std::move(core_filter));
+
         signal_dispatcher->on<io::context::os_signal>([=, &context](int signum, siginfo_t){
             if(signum == SIGHUP) {
                 COCAINE_LOG_INFO(internal_logger, "resetting logging v2 root logger");
                 *root_logger = get_root_logger(context, config);
             }
         });
         context.signal_hub().listen(signal_dispatcher, io_context);
-        init();
     }
 
-    void on_filter_list(std::future<response::children_subscribe> future) {
-        try {
-            auto filter_ids = future.get();
-            COCAINE_LOG_DEBUG(internal_logger,
-                              "received filter list update, count - {}",
-                              std::get<1>(filter_ids).size());
-            for (auto& id_str : std::get<1>(filter_ids)) {
-                logging::filter_t::id_type filter_id = std::stoull(id_str);
-                listen_for_filter(filter_id);
+    template <class T, class F>
+    struct
+    safe_t {
+        safe_t(std::weak_ptr<T> _self, F _f):
+            self(std::move(_self)),
+            f(std::move(_f))
+        {}
+
+        template<class... Args>
+        auto operator()(Args&&... args) -> void {
+            if(auto lock = self.lock()) {
+                f(std::forward<Args>(args)...);
```

Comment:
As discussed - we can not control circular dependency here and callback invocation, so yes, in case of service shutdown those callbacks will be silently ignored - we are not interested in them.